client/auth.go

```diff
@@ -64,50 +74,95 @@ func (c *Conn) readInitialHandshake() error {
 		// mysql-proxy also use 12
 		// which is not documented but seems to work.
 		c.salt = append(c.salt, data[pos:pos+12]...)
+		pos += 13
+		// auth plugin
+		if end := bytes.IndexByte(data[pos:], 0x00); end != -1 {
+			c.authPluginName = string(data[pos : pos+end])
+		} else {
+			c.authPluginName = string(data[pos:])
+		}
+	}
+
+	// if server gives no default auth plugin name, use a client default
+	if c.authPluginName == "" {
+		c.authPluginName = defaultAuthPluginName
 	}
 
 	return nil
 }
 
+// generate auth response data according to auth plugin
+func (c *Conn) genAuthResponse(authData []byte) ([]byte, bool, error) {
+	// password hashing
+	switch c.authPluginName {
+	case "mysql_native_password":
+		return CalcPassword(authData[:20], []byte(c.password)), false, nil
+	case "caching_sha2_password":
+		return CalcCachingSha2Password(authData, c.password), false, nil
+	case "sha256_password":
+		if len(c.password) == 0 {
+			return nil, true, nil
+		}
+		if c.tlsConfig != nil || c.proto == "unix" {
+			// write cleartext auth packet
+			return []byte(c.password), true, nil
+		} else {
+			// request public key from server
+			// see: https://dev.mysql.com/doc/internals/en/public-key-retrieval.html
+			return []byte{1}, false, nil
+		}
+	default:
+		// not reachable
+		return nil, false, fmt.Errorf("auth plugin '%s' is not supported", c.authPluginName)
+	}
+}
+
+// See: http://dev.mysql.com/doc/internals/en/connection-phase-packets.html#packet-Protocol::HandshakeResponse
 func (c *Conn) writeAuthHandshake() error {
+	if c.authPluginName != "mysql_native_password" && c.authPluginName != "caching_sha2_password" && c.authPluginName != "sha256_password" {
```

Comment:
can you extract to a function like `authPluginAllowed(name)`

Reply:
sure